exp/query/cache.go

```diff
@@ -0,0 +1,158 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package query
+
+import (
+	"container/list"
+	"errors"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+)
+
+var (
+	ErrTenantNotFound       = errors.New("cache: tenant not found")
+	ErrTenantDirectoryFound = errors.New("cache: tenant directory not found")
+)
+
+const (
+	evictFraction = 0.1 // When evicting, remove 10% of max disk capacity.
+	CachePrefix   = "weaviate-cache"
+)
+
+// DiskCache is LRU disk based cache for tenants on-disk data.
+type DiskCache struct {
+	// key: <collection>/<tenant>
+	tenants   map[string]*list.Element
+	evictList *list.List
+	mu        sync.Mutex // TODO(kavi): Make it more fine-grained than global mutex
+
+	// more than this will start evicting the tenants
+	// maxCap in bytes.
+	maxCap int64
+
+	basePath string
+}
+
+func NewDiskCache(basePath string, maxCap int64) *DiskCache {
+	return &DiskCache{
+		tenants:   make(map[string]*list.Element),
+		evictList: list.New(),
+		basePath:  path.Join(basePath, CachePrefix),
+		maxCap:    maxCap,
+	}
+}
+
+// Assumes id all small-case
+func (d *DiskCache) Tenant(collection, tenantID string) (*TenantCache, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if c, exists := d.tenants[d.TenantKey(collection, tenantID)]; exists {
+		d.evictList.MoveToFront(c)
+		c.Value.(*TenantCache).LastAccessed = time.Now()
+		return c.Value.(*TenantCache), nil
+	}
+	return nil, ErrTenantNotFound
+}
+
+// AddTenant is called after having the files in right directory
+// AddTenant checks if file is present on that directory
+// deterministically generated from `collection` & `tenant` with basePath
+func (d *DiskCache) AddTenant(collection, tenantID string, version uint64) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	tc := &TenantCache{
+		TenantID:     tenantID,
+		Collection:   collection,
+		Version:      version,
+		LastAccessed: time.Now(),
+		basePath:     d.basePath,
+	}
+
+	_, err := os.Stat(tc.AbsolutePath())
+	if errors.Is(err, os.ErrNotExist) {
+		return errors.Join(ErrTenantDirectoryFound, err)
+	}
+
+	c := d.evictList.PushFront(tc)
+	d.tenants[d.TenantKey(tc.Collection, tc.TenantID)] = c
+
+	// evit if the size is getting filled up
+	// TODO(kavi): Worth doing it on different go routine? But will get complex with correctness
+	if d.usage() > d.maxCap {
+		if err := d.evict(); err != nil {
```

Comment:
What happens if we evict a tenant in parallel while that tenant is being searched or downloaded to? I think we can come back to this later, but should leave a TODO in code or create an issue to handle it

Reply:
> if we evict a tenant in parallel while that tenant is being searched or downloaded to?
This case happens, only say a particular tenant is not being used for long time (till eviction kicked in) and this tenant reached the end of eviction list and suddenly the request for this tenant comes in. What happens here is, if eviction wins the race, we will do one extra upstream call.

This is the inherent problem with LRU eviction policy. I don't think any sane solution for it unless I miss something.

Reply:
Could we mark tenant for "eviction" and when a search lands we ensure that we're not using a tenant marked for eviction ? If we have no alternative version we can "unmark" it for eviction ? 

Just a suggestion and can come as a follow up PR

Reply:
explained why it's hard to do that [here](https://github.com/weaviate/weaviate/pull/6267#discussion_r1836713824) :) Happy to discuss if i'm missing anything.